### ISSUE TITLE
Unicode error with certain episodes when syncing to device

### ISSUE THREAD

**Issue by alexweissman** [gpodder/gpodder#139]
I get the following error in my log when trying to synchronize certain podcasts with my device:

```
2016-07-30 14:13:14,355 [gpodder.log] ERROR: Uncaught exception: Traceback (most recent call last):
  File "C:\Program Files (x86)\gPodder\src\gpodder\gtkui\desktop\sync.py", line 282, in check_free_space
    resume_sync([],[],None)
  File "C:\Program Files (x86)\gPodder\src\gpodder\gtkui\desktop\sync.py", line 193, in resume_sync
    playlist.write_m3u(episodes_for_playlist)
  File "C:\Program Files (x86)\gPodder\src\gpodder\gtkui\desktop\deviceplaylist.py", line 110, in write_m3u
    filename = self.get_absolute_filename_for_playlist(current_episode)
  File "C:\Program Files (x86)\gPodder\src\gpodder\gtkui\desktop\deviceplaylist.py", line 89, in get_absolute_filename_for_playlist
    filename = os.path.join(util.relpath(self.mountpoint, self._config.device_sync.device_folder), filename)
  File "C:\Python27\Lib\ntpath.py", line 84, in join
    result_path = result_path + p_path
UnicodeDecodeError: 'utf8' codec can't decode byte 0x92 in position 26: invalid start byte
```

According to [this answer](http://stackoverflow.com/a/29419477/2970321), `0x92` is a Windows "smart quote", and I do see quite a few smart quotes scattered throughout various episode titles.

When I throw some debugging statements into `deviceplaylist.py`, it does indeed to be halting on episodes that contain apostrophes and quotes.

**Comment by alexweissman** [comment]
I'm guessing this happens whenever the person responsible for generating the feed at the podcast copy-pastes the filename from a Word doc that someone gave them ;-)

**Comment by alexweissman** [comment]
Setting the default value of `use_ascii` in [sanitize_filename](https://github.com/gpodder/gpodder/blob/master/src/gpodder/util.py#L1306) to `True` seems to work as a dirty fix, since any filenames involved in a one-way sync have to go through that function.

It looks like `sanitize_filename` is actually supposed to detect the encoding and automatically convert it to something usable, but it seems like it is incorrectly identifying everything as `MBCS` instead of what are actually `windows-1252` filenames.

**Comment by thp** [comment]
"on Windows, Python uses the name “mbcs” to refer to whatever the currently configured encoding is" (from https://docs.python.org/2/howto/unicode.html)

Not sure why it tries UTF-8 in your traceback above, though..

**Comment by thp** [comment]
Do you have an example feed URL (and episode name) for which this happens, so I can try to reproduce this issue locally?

**Comment by alexweissman** [comment]
Yup, try the 99% Invisible feed: http://feeds.99percentinvisible.org/99percentinvisible, with episode "America's Last Top Model".  The `'` in "America's" is one of the `windows-1252` characters.

**Comment by thp** [comment]
Does the device you try to sync to have FAT32 as filesystem?
I couldn't reproduce it on an NTFS filesystem and do not have anything with a FAT32 system here at the moment in this Windows VM.

**Comment by alexweissman** [comment]
Good question.  Yes, it is a FAT32 filesystem on my device - specifically, the Sandisk Clip+ with a 32GB removable memory card installed and running [Rockbox](http://rockbox.org/).

**Comment by thp** [comment]
I've tried reproducing it on a Windows 10 installation with "German (Austria)" as locale. With gPodder 3.9.1, syncing to a FAT32 USB disk, and Python 2.7.12, I get the file to properly sync. Also the M3U file is properly created (as latin-1).

Can you try the following?

Before the line:

```
filename = os.path.join(util.relpath(self.mountpoint, self._config.device_sync.device_folder), filename)
```

Add these prints:

```
print 'A:', repr(self.mountpoint)
print 'B:', repr(self._config.device_sync.device_folder)
print 'C:', repr(filename)
print 'D:', repr(util.relpath(self.mountpoint, self._config.device_sync.device_folder))
```

Please report back what your findings are. I've tried a few combinations of passing parameters to `os.path.join()`, but couldn't reproduce this exact error. I did notice that the path listed in the m3u files is wrong for me, so maybe the whole playlist creation code is kinda broken at the moment?

**Comment by thp** [comment]
@alexweissman Have you been able to test the things suggested in my last post?

**Comment by alexweissman** [comment]
I just installed gPodder 3.9.1 (in a separate directory, I'm not using version control here and I didn't want to overwrite my custom hacks to 3.9.0.  I should probably just check out the repo but I'm too lazy to set up git on my Windows machine, but that's another story).

3.9.1 seems to be pulling my configuration and database correctly from the 3.9.0 install, but now when I try to synchronize, I immediately get the error:

> Error writing playlist files:
> [Errornum 9] Bad file descriptor

It does not appear to write anything to the log.

I'm signed into the #gpodder IRC channel, BTW, if you want to have an interactive discussion about this :)

**Comment by elelay** [comment]
Closing since it's really and might be fixed in the 3.10 branch with the switch to python 3.
